openedx_lti_tool_plugin/deep_linking/forms.py

```diff
@@ -1,131 +1,208 @@
+
+        Example:
+            OLTITP_DL_CONTENT_ITEMS_PROVIDER = 'example.module.path.provider_function'
+
+        This method will then try to call the provider function, the call will include
+        the HTTPRequest object and deep linking launch data dictionary received from
+        the deep linking request has arguments.
+
+        The content items returned from the provider function must be a list of dictionaries,
+        this list will be validated with a JSON schema validator using a JSON schema defined
+        in the CONTENT_ITEMS_SCHEMA constant.
 
         Returns:
-            Tuple containing the choice value and name.
+            A list with content item dictionaries.
+
+            An empty list if OLTITP_DL_CONTENT_ITEMS_PROVIDER setting is None.
+            or there was an Exception importing or calling the provider function,
+            or the data returned by the provider function is not valid.
+            or the provider function returned an empty list.
 
-        .. _LTI Deep Linking Specification - LTI Resource Link:
-            https://www.imsglobal.org/spec/lti-dl/v2p0#lti-resource-link
+        .. _LTI Deep Linking Specification - Content Item Types:
+            https://www.imsglobal.org/spec/lti-dl/v2p0#content-item-types
 
         """
-        relative_url = reverse(
-            f'{app_config.name}:1.3:resource-link:launch-course',
-            kwargs={'course_id': course.learning_context.context_key},
-        )
+        if not (setting := configuration_helpers().get_value(
+            'OLTITP_DL_CONTENT_ITEMS_PROVIDER',
+            settings.OLTITP_DL_CONTENT_ITEMS_PROVIDER,
+        )):
+            return []
 
-        return (
-            self.request.build_absolute_uri(relative_url),
-            course.learning_context.title,
-        )
+        try:
+            path, name = str(setting).rsplit('.', 1)
+            content_items = getattr(import_module(path), name)(
+                self.request,
+                self.launch_data,
+            )
+            validate(content_items, self.CONTENT_ITEMS_SCHEMA)
 
-    def get_course_contexts(self):
-        """Get CourseContext objects.
+            return content_items
+        except Exception as exc:  # pylint: disable=broad-exception-caught
+            log_extra = {
+                'setting': setting,
+                'exception': str(exc),
+            }
+            log.error(f'Error obtaining content items from provider: {log_extra}')
 
-        Returns:self.cleaned_data
-            All CourseContext objects if COURSE_ACCESS_CONFIGURATION switch
-            is disabled or all CourseContext objects matching the IDs in
-            the CourseAccessConfiguration `allowed_course_ids` field.
+            return []
 
-        Raises:
-            CourseAccessConfiguration.DoesNotExist: If CourseAccessConfiguration
-            does not exist for this form `lti_tool` attribute.
+    def get_content_items(self) -> List[Optional[dict]]:
```

Comment:
What do you think of specifying, on the name of the function, that it gets the content items from the course access configuration allowed course IDs? With the current name I'd expect the function to get the content items in general.
```suggestion
    def get_content_items_course_access(self) -> List[Optional[dict]]:
```

Reply:
@alexjmpb This method obtains the content items in general unless a content items provider is setup, these could also include in the future obtaining the content items for course units or course problems/components, I would not think that renaming the method `get_content_items_course_access` will be appropriate because if the course access configuration switch is disabled this method will not filter the content items using the `CourseAccessConfiguration.allowed_course_ids` value. What do you think if instead we rename it something like `get_content_items_from_lms` or `get_content_items_from_edxapp`?